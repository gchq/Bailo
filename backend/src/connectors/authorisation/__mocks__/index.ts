import { vi } from 'vitest'

const auth = {
  hasModelVisibilityAccess: vi.fn(() => true),
  hasAccessRequest: vi.fn(() => true),

  model: vi.fn(() => ({ success: true })),
  models: vi.fn(() => [{ success: true }]),

  schema: vi.fn(() => ({ success: true })),
  schemas: vi.fn(() => [{ success: true }]),

  schemaMigration: vi.fn(() => ({ success: true })),
  schemaMigrations: vi.fn(() => ({ success: true })),

  accessRequest: vi.fn(() => ({ success: true })),
  accessRequests: vi.fn(() => [{ success: true }]),

  release: vi.fn(() => ({ success: true })),
  releases: vi.fn(() => [{ success: true }]),

  file: vi.fn(() => ({ success: true })),
  files: vi.fn(() => [{ success: true }]),

  image: vi.fn(() => ({ success: true })),
  images: vi.fn(() => [{ success: true }]),
}

export default auth
