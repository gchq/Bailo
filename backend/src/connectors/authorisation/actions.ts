import { TokenActions } from '../../models/Token.js'

export const ModelAction = {
  Create: 'model:create',
  View: 'model:view',
  Update: 'model:update',
  Write: 'model:write',
} as const
export type ModelActionKeys = (typeof ModelAction)[keyof typeof ModelAction]

export const ReleaseAction = {
  Create: 'release:create',
  View: 'release:view',
  Delete: 'release:delete',
  Update: 'release:update',
} as const
export type ReleaseActionKeys = (typeof ReleaseAction)[keyof typeof ReleaseAction]

export const AccessRequestAction = {
  Create: 'access_request:create',
  View: 'access_request:view',
  Update: 'access_request:update',
  Delete: 'access_request:delete',
} as const
export type AccessRequestActionKeys = (typeof AccessRequestAction)[keyof typeof AccessRequestAction]

export const SchemaAction = {
  Create: 'schema:create',
  Delete: 'schema:delete',
  Update: 'schema:update',
} as const
export type SchemaActionKeys = (typeof SchemaAction)[keyof typeof SchemaAction]

export const FileAction = {
  Delete: 'file:delete',
  Upload: 'file:upload',
  // 'view' refers to the ability to see metadata about the file.  'download' lets the user view the file contents.
  View: 'file:view',
  Download: 'file:download',
  Update: 'file:update',
} as const
export type FileActionKeys = (typeof FileAction)[keyof typeof FileAction]

export const ImageAction = {
  Pull: 'image:pull',
  Push: 'image:push',
  List: 'image:list',
  Wildcard: 'image:wildcard',
  Delete: 'image:delete',
} as const
export type ImageActionKeys = (typeof ImageAction)[keyof typeof ImageAction]

export const ResponseAction = {
  Create: 'response:create',
  View: 'response:view',
  Update: 'response:update',
} as const
export type ResponseActionKeys = (typeof ResponseAction)[keyof typeof ResponseAction]

export const ActionLookup = {
  [ModelAction.Create]: TokenActions.ModelWrite.id,
  [ModelAction.View]: TokenActions.ModelRead.id,
  [ModelAction.Update]: TokenActions.ModelWrite.id,
  [ModelAction.Write]: TokenActions.ModelWrite.id,

  [ReleaseAction.Create]: TokenActions.ReleaseWrite.id,
  [ReleaseAction.View]: TokenActions.ReleaseRead.id,
  [ReleaseAction.Delete]: TokenActions.ReleaseWrite.id,
  [ReleaseAction.Update]: TokenActions.ReleaseWrite.id,

  [AccessRequestAction.Create]: TokenActions.AccessRequestWrite.id,
  [AccessRequestAction.View]: TokenActions.AccessRequestRead.id,
  [AccessRequestAction.Update]: TokenActions.AccessRequestWrite.id,
  [AccessRequestAction.Delete]: TokenActions.AccessRequestWrite.id,

  [SchemaAction.Create]: TokenActions.SchemaWrite.id,
  [SchemaAction.Delete]: TokenActions.SchemaWrite.id,
  [SchemaAction.Update]: TokenActions.SchemaWrite.id,

  [FileAction.Delete]: TokenActions.FileWrite.id,
  [FileAction.Upload]: TokenActions.FileWrite.id,
  [FileAction.View]: TokenActions.FileRead.id,
  [FileAction.Download]: TokenActions.FileRead.id,
  [FileAction.Update]: TokenActions.FileWrite.id,

  [ImageAction.Pull]: TokenActions.ImageRead.id,
  [ImageAction.Push]: TokenActions.ImageWrite.id,
  [ImageAction.List]: TokenActions.ImageRead.id,
  [ImageAction.Wildcard]: TokenActions.ImageWrite.id,
  [ImageAction.Delete]: TokenActions.ImageWrite.id,
} as const
export type ActionLookupKeys = (typeof ActionLookup)[keyof typeof ActionLookup]
