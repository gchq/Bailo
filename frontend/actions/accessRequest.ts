import useSWR from 'swr'
import { AccessRequestInterface, AccessRequestUserPermissions } from 'types/types'
import { ErrorInfo, fetcher } from 'utils/fetcher'

const emptyAccessRequestList = []

export function useGetAccessRequestsForModelId(modelId?: string) {
  const { data, isLoading, error, mutate } = useSWR<
    {
      accessRequests: AccessRequestInterface[]
    },
    ErrorInfo
  >(modelId ? `/api/v2/model/${modelId}/access-requests` : null, fetcher)

  return {
    mutateAccessRequests: mutate,
    accessRequests: data ? data.accessRequests : emptyAccessRequestList,
    isAccessRequestsLoading: isLoading,
    isAccessRequestsError: error,
  }
}

export function useGetAccessRequest(modelId: string | undefined, accessRequestId: string | undefined) {
  const { data, isLoading, error, mutate } = useSWR<
    {
      accessRequest: AccessRequestInterface
    },
    ErrorInfo
  >(accessRequestId && modelId ? `/api/v2/model/${modelId}/access-request/${accessRequestId}` : null, fetcher)

  return {
    mutateAccessRequest: mutate,
    accessRequest: data?.accessRequest,
    isAccessRequestLoading: isLoading,
    isAccessRequestError: error,
  }
}

export function useGetCurrentUserPermissionsForAccessRequest(entryId?: string, accessRequestId?: string) {
  const { data, isLoading, error, mutate } = useSWR<
    {
      permissions: AccessRequestUserPermissions
    },
    ErrorInfo
  >(
    entryId && accessRequestId ? `/api/v2/model/${entryId}/access-request/${accessRequestId}/permissions/mine` : null,
    fetcher,
  )

  return {
    mutateAccessRequestUserPermissions: mutate,
    accessRequestUserPermissions: data?.permissions,
    isAccessRequestUserPermissionsLoading: isLoading,
    isAccessRequestUserPermissionsError: error,
  }
}

export function postAccessRequest(modelId: string, schemaId: string, form: Record<string, unknown>) {
  return fetch(`/api/v2/model/${modelId}/access-requests`, {
    method: 'post',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ metadata: { ...form }, schemaId: schemaId }),
  })
}

export function deleteAccessRequest(modelId: string, accessRequestId: string) {
  return fetch(`/api/v2/model/${modelId}/access-request/${accessRequestId}`, {
    method: 'delete',
    headers: { 'Content-Type': 'application/json' },
  })
}

export function patchAccessRequest(modelId: string, accessRequestId: string, form: Record<string, unknown>) {
  return fetch(`/api/v2/model/${modelId}/access-request/${accessRequestId}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ metadata: { ...form } }),
  })
}

export function postAccessRequestComment(modelId: string, accessRequestId: string, comment: string) {
  return fetch(`/api/v2/model/${modelId}/access-request/${accessRequestId}/comment`, {
    method: 'post',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ comment }),
  })
}
