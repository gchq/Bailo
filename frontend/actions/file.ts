import axios, { AxiosProgressEvent } from 'axios'
import qs from 'querystring'
import useSWR from 'swr'
import { FileInterface, FileWithScanResultsInterface } from 'types/types'
import { handleAxiosError } from 'utils/axios'
import { ErrorInfo, fetcher } from 'utils/fetcher'

const emptyFilesList = []

export function useGetFilesForModel(modelId: string) {
  const { data, isLoading, error, mutate } = useSWR<
    {
      files: FileWithScanResultsInterface[]
    },
    ErrorInfo
  >(`/api/v2/model/${modelId}/files`, fetcher)

  return {
    mutateFiles: mutate,
    files: data ? data.files : emptyFilesList,
    isFilesLoading: isLoading,
    isFilesError: error,
  }
}

export async function postFileForModelId(
  modelId: string,
  file: File,
  onUploadProgress: (progress: AxiosProgressEvent) => void,
  metadata?: string,
  tags: string[] = [],
) {
  const mime = file.type || 'application/octet-stream'

  const queryParams = {
    ...(metadata && { metadata }),
    ...(tags.length > 0 && { tags }),
  }
  const fileResponse = await axios
    .post(
      `/api/v2/model/${modelId}/files/upload/simple?name=${file.name}&mime=${mime}&${qs.stringify(queryParams)}}`,
      file,
      {
        onUploadProgress,
      },
    )
    .catch(function (error) {
      if (error.response) {
        throw new Error(
          `Error code ${error.response.status} received from server whilst attempting to upload file ${file.name}`,
        )
      } else if (error.request) {
        throw new Error(`There was a problem with the request whilst attempting to upload file ${file.name}`)
      } else {
        throw new Error(`Unknown error whilst attempting to upload file ${file.name}`)
      }
    })
  return fileResponse
}

export async function patchFile(modelId: string, fileId: string, metadata: Pick<FileInterface, 'tags'>) {
  try {
    const response = await axios({
      method: 'patch',
      url: `/api/v2/model/${modelId}/file/${fileId}`,
      headers: { 'Content-Type': 'application/json' },
      data: { tags: metadata.tags },
    })
    return { status: response.status, data: response.data }
  } catch (error) {
    return handleAxiosError(error)
  }
}
