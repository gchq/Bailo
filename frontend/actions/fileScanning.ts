import useSWR from 'swr'
import { ErrorInfo, fetcher } from 'utils/fetcher'

const emptyScannerList = []

export function useGetFileScannerInfo() {
  const { data, isLoading, error, mutate } = useSWR<
    {
      scanners: string[]
    },
    ErrorInfo
  >('/api/v2/filescanning/info', fetcher)

  return {
    scannersMutate: mutate,
    scanners: data ? data.scanners : emptyScannerList,
    isScannersLoading: isLoading,
    isScannersError: error,
  }
}

export function rerunFileScan(modelId: string, fileId: string) {
  return fetch(`/api/v2/filescanning/model/${modelId}/file/${fileId}/scan`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
  })
}
