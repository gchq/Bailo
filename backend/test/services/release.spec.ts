import { describe, expect, test, vi } from 'vitest'

import { ReleaseAction } from '../../src/connectors/authorisation/actions.js'
import authorisation from '../../src/connectors/authorisation/index.js'
import {
  createRelease,
  deleteRelease,
  getAllFileIds,
  getFileByReleaseFileName,
  getModelReleases,
  getReleaseBySemver,
  getReleasesForExport,
  newReleaseComment,
  removeFileFromReleases,
  updateRelease,
} from '../../src/services/release.js'

vi.mock('../../src/connectors/authorisation/index.js')

const responseModelMock = vi.hoisted(() => {
  const obj: any = {}

  obj.aggregate = vi.fn(() => obj)
  obj.match = vi.fn(() => obj)
  obj.sort = vi.fn(() => obj)
  obj.lookup = vi.fn(() => obj)
  obj.append = vi.fn(() => obj)
  obj.find = vi.fn(() => obj)
  obj.findOne = vi.fn(() => obj)
  obj.findOneAndUpdate = vi.fn(() => obj)
  obj.findByIdAndUpdate = vi.fn(() => obj)
  obj.updateOne = vi.fn(() => obj)
  obj.save = vi.fn(() => obj)
  obj.limit = vi.fn(() => obj)
  obj.unwind = vi.fn(() => obj)
  obj.at = vi.fn(() => obj)
  obj.map = vi.fn(() => [])
  obj.filter = vi.fn(() => [])

  const model: any = vi.fn(() => obj)
  Object.assign(model, obj)

  return model
})

vi.mock('../../src/models/Response.js', async () => ({
  ...((await vi.importActual('../../src/models/Response.js')) as object),
  default: responseModelMock,
}))

const modelMocks = vi.hoisted(() => ({
  getModelById: vi.fn(() => ({
    id: 'test_model_id',
    card: { version: 1 },
    settings: { mirror: { sourceModelId: '' } },
  })),
  getModelCardRevision: vi.fn(),
}))
vi.mock('../../src/services/model.js', () => modelMocks)

const registryMocks = vi.hoisted(() => ({
  listModelImages: vi.fn(),
}))
vi.mock('../../src/services/registry.js', () => registryMocks)

const fileMocks = vi.hoisted(() => ({
  getFileById: vi.fn(),
  getFilesByIds: vi.fn(),
}))
vi.mock('../../src/services/file.js', () => fileMocks)

const releaseModelMocks = vi.hoisted(() => {
  const obj: any = {}

  obj.aggregate = vi.fn(() => obj)
  obj.match = vi.fn(() => obj)
  obj.unwind = vi.fn(() => obj)
  obj.group = vi.fn(() => obj)
  obj.sort = vi.fn(() => obj)
  obj.lookup = vi.fn(() => obj)
  obj.append = vi.fn(() => obj)
  obj.find = vi.fn(() => obj)
  obj.findOne = vi.fn(() => obj)
  obj.updateOne = vi.fn(() => obj)
  obj.updateMany = vi.fn(() => obj)
  obj.save = vi.fn(() => obj)
  obj.delete = vi.fn(() => obj)
  obj.findOneAndUpdate = vi.fn(() => obj)
  obj.findOneWithDeleted = vi.fn(() => obj)
  obj.filter = vi.fn(() => obj)

  const model: any = vi.fn((params) => ({ ...obj, ...params }))
  Object.assign(model, obj)

  return model
})
vi.mock('../../src/models/Release.js', () => ({ default: releaseModelMocks }))

const mockReviewService = vi.hoisted(() => {
  return {
    createReleaseReviews: vi.fn(),
  }
})
vi.mock('../../src/services/review.js', () => mockReviewService)

const mockWebhookService = vi.hoisted(() => {
  return {
    sendWebhooks: vi.fn(),
  }
})
vi.mock('../../src/services/webhook.js', () => mockWebhookService)

describe('services > release', () => {
  test('createRelease > simple', async () => {
    releaseModelMocks.findOneWithDeleted.mockResolvedValue(null)

    await createRelease({} as any, { semver: 'v1.0.0', minor: false } as any)

    expect(releaseModelMocks.save).toBeCalled()
    expect(releaseModelMocks).toBeCalled()
    expect(mockReviewService.createReleaseReviews).toBeCalled()
    expect(mockWebhookService.sendWebhooks).toBeCalled()
  })

  test('createRelease > minor release', async () => {
    releaseModelMocks.findOneWithDeleted.mockResolvedValue(null)

    await createRelease({} as any, { semver: 'v1.0.0', minor: true } as any)

    expect(releaseModelMocks.save).toBeCalled()
    expect(releaseModelMocks).toBeCalled()
    expect(mockReviewService.createReleaseReviews).not.toBeCalled()
  })

  test('createRelease > release with image', async () => {
    const existingImages = [{ repository: 'mockRep', name: 'image', tags: ['latest'] }]
    registryMocks.listModelImages.mockResolvedValueOnce(existingImages)

    releaseModelMocks.findOneWithDeleted.mockResolvedValue(null)

    await createRelease(
      {} as any,
      {
        semver: 'v1.0.0',
        images: existingImages.flatMap(({ tags, ...rest }) => tags.map((tag) => ({ tag, ...rest }))),
      } as any,
    )

    expect(releaseModelMocks.save).toBeCalled()
    expect(releaseModelMocks).toBeCalled()
    expect(mockReviewService.createReleaseReviews).toBeCalled()
  })

  test('createRelease > missing images in the registry', async () => {
    const existingImages = [{ repository: 'mockRep', name: 'image', tags: ['latest'] }]
    registryMocks.listModelImages.mockResolvedValueOnce(existingImages)
    releaseModelMocks.findOneWithDeleted.mockResolvedValue(null)

    expect(() =>
      createRelease(
        {} as any,
        {
          semver: 'v1.0.0',
          modelCardVersion: 999,
          images: [
            { repository: 'fake', name: 'fake', tag: 'fake1' },
            { repository: 'fake', name: 'fake', tag: 'fake2' },
          ].concat(existingImages.flatMap(({ tags, ...rest }) => tags.map((tag) => ({ tag, ...rest })))),
        } as any,
      ),
    ).rejects.toThrowError(/^The following images do not exist in the registry/)
    expect(releaseModelMocks.save).not.toBeCalled()
    expect(releaseModelMocks).not.toBeCalled()
    expect(mockReviewService.createReleaseReviews).not.toBeCalled()
  })

  test('createRelease > release with bad files', async () => {
    fileMocks.getFileById.mockResolvedValueOnce({ modelId: 'random_model' })
    releaseModelMocks.findOneWithDeleted.mockResolvedValue(null)

    expect(() =>
      createRelease(
        {} as any,
        {
          semver: 'v1.0.0',
          modelCardVersion: 999,
          fileIds: ['test'],
        } as any,
      ),
    ).rejects.toThrowError(/^The file 'test' comes from the model/)

    expect(releaseModelMocks.save).not.toBeCalled()
  })

  test('createRelease > release with duplicate file names', async () => {
    fileMocks.getFileById.mockResolvedValue({ modelId: 'test_model_id', name: 'test_file.png' })
    modelMocks.getModelById.mockResolvedValue({
      id: 'test_model_id',
      card: { version: 1 },
      settings: { mirror: { sourceModelId: '' } },
    })
    releaseModelMocks.findOneWithDeleted.mockResolvedValue(null)

    expect(
      async () =>
        await createRelease(
          {} as any,
          {
            semver: 'v1.0.0',
            modelCardVersion: 999,
            fileIds: ['test', 'test2'],
          } as any,
        ),
    ).rejects.toThrowError(/^Releases cannot have multiple files with the same name/)

    expect(releaseModelMocks.save).not.toBeCalled()
  })

  test('createRelease > release with bad semver', async () => {
    releaseModelMocks.findOneWithDeleted.mockResolvedValue(null)
    const result = createRelease(
      {} as any,
      {
        semver: 'bad semver',
        modelCardVersion: 999,
      } as any,
    )
    expect(result).rejects.toThrowError(/is not a valid semver value./)

    expect(releaseModelMocks.save).not.toBeCalled()
  })

  test('createRelease > bad authorisation', async () => {
    vi.mocked(authorisation.release).mockResolvedValue({ info: 'You do not have permission', success: false, id: '' })

    releaseModelMocks.findOneWithDeleted.mockResolvedValue(null)
    expect(() => createRelease({} as any, { semver: 'v1.0.0' } as any)).rejects.toThrowError(
      /^You do not have permission/,
    )
  })

  test('createRelease > automatic model card version', async () => {
    modelMocks.getModelById.mockResolvedValueOnce({
      id: 'test_model_id',
      card: { version: 999 },
      settings: { mirror: { sourceModelId: '' } },
    })
    releaseModelMocks.findOneWithDeleted.mockResolvedValue(null)

    await createRelease({} as any, { semver: 'v1.0.0' } as any)

    expect(releaseModelMocks.save).toBeCalled()
    expect(releaseModelMocks.mock.calls.at(0)[0].modelCardVersion).toBe(999)
  })

  test('createRelease > no model card', async () => {
    modelMocks.getModelById.mockResolvedValueOnce({
      id: 'test_model_id',
      card: undefined as any,
      settings: { mirror: { sourceModelId: '' } },
    })

    expect(() => createRelease({} as any, { semver: 'v1.0.0' } as any)).rejects.toThrowError(
      /^This model does not have a model card associated with it/,
    )

    expect(releaseModelMocks.save).not.toBeCalled()
  })

  test('createRelease > should throw Bad Req if the user tries to alter a mirrored model card', async () => {
    vi.mocked(authorisation.release).mockResolvedValueOnce({
      info: 'Cannot create a release from a mirrored model',
      success: false,
      id: '',
    })
    releaseModelMocks.findOneWithDeleted.mockResolvedValue(null)
    expect(() => createRelease({} as any, { semver: 'v1.0.0' } as any)).rejects.toThrowError(
      /^Cannot create a release from a mirrored model/,
    )
  })

  test('updateRelease > bad authorisation', async () => {
    vi.mocked(authorisation.release).mockResolvedValue({ info: 'You do not have permission', success: false, id: '' })
    expect(() => updateRelease({} as any, 'model-id', 'v1.0.0', {} as any)).rejects.toThrowError(
      /^You do not have permission/,
    )
  })

  test('updateRelease > release with bad files', async () => {
    fileMocks.getFileById.mockResolvedValueOnce({ modelId: 'random_model' })

    expect(() =>
      updateRelease({} as any, 'model-id', 'v1.0.0', {
        semver: 'v1.0.0',
        fileIds: ['test'],
      } as any),
    ).rejects.toThrowError(/^The file 'test' comes from the model/)
  })

  test('updateRelease > success', async () => {
    releaseModelMocks.findOne.mockResolvedValue({ semver: 'v1.0.0' })

    await updateRelease({} as any, 'model-id', 'v1.0.0', { notes: 'New notes' } as any)

    expect(releaseModelMocks.findOneAndUpdate).toBeCalled()
  })

  test('updateRelease > should throw Bad Req when attempting to update a release on a mirrored model ', async () => {
    vi.mocked(authorisation.release).mockResolvedValue({
      info: 'Cannot update a release on a mirrored model.',
      success: false,
      id: '',
    })
    expect(() => updateRelease({} as any, 'model-id', 'v1.0.0', {} as any)).rejects.toThrowError(
      /^Cannot update a release on a mirrored model./,
    )
  })

  test('newReleaseComment > success', async () => {
    releaseModelMocks.findOneAndUpdate.mockResolvedValue({})

    await newReleaseComment({} as any, 'model', '1.0.0', 'This is a new comment')

    expect(responseModelMock.save).toBeCalled()
  })

  test('newReleaseComment > should throw bad request when attempting to create a release comment on a mirrored model', async () => {
    vi.mocked(authorisation.release).mockResolvedValue({
      info: 'Cannot create a new comment on a mirrored model.',
      success: false,
      id: '',
    })
    modelMocks.getModelById.mockResolvedValueOnce({
      id: 'test_model_id',
      card: { version: 1 },
      settings: { mirror: { sourceModelId: '123' } },
    })
    expect(() => newReleaseComment({} as any, 'model', '1.0.0', 'This is a new comment')).rejects.toThrowError(
      /^Cannot create a new comment on a mirrored model./,
    )
    expect(releaseModelMocks.findOneAndUpdate).not.toBeCalled()
  })

  test('getModelReleases > good', async () => {
    await getModelReleases({} as any, 'modelId')

    vi.mocked(releaseModelMocks.lookup).mockImplementation(() => ({
      ...releaseModelMocks.lookup,
    }))

    expect(releaseModelMocks.match.mock.calls.at(0)).toMatchSnapshot()
    expect(releaseModelMocks.sort.mock.calls.at(0)).toMatchSnapshot()
    expect(releaseModelMocks.lookup.mock.calls.at(0)).toMatchSnapshot()
    expect(releaseModelMocks.append.mock.calls.at(0)).toMatchSnapshot()
  })

  test('getReleaseBySemver > good', async () => {
    const mockRelease = { _id: 'release' }

    releaseModelMocks.findOne.mockResolvedValue(mockRelease)

    expect(await getReleaseBySemver({} as any, 'test', 'test')).toBe(mockRelease)
  })

  test('getReleaseBySemver > no release', async () => {
    releaseModelMocks.findOne.mockResolvedValue(undefined)

    expect(() => getReleaseBySemver({} as any, 'test', 'test')).rejects.toThrowError(
      /^The requested release was not found./,
    )
  })

  test('getReleaseBySemver > no permission', async () => {
    const mockRelease = { _id: 'release' }

    releaseModelMocks.findOne.mockResolvedValue(mockRelease)
    vi.mocked(authorisation.release).mockResolvedValue({
      info: 'You do not have permission to view this release.',
      success: false,
      id: '',
    })

    expect(() => getReleaseBySemver({} as any, 'test', 'test')).rejects.toThrowError(
      /^You do not have permission to view this release./,
    )
  })

  test('deleteRelease > success', async () => {
    expect(await deleteRelease({} as any, 'test', 'test')).toStrictEqual({ modelId: 'test', semver: 'test' })
  })

  test('deleteRelease > no permission', async () => {
    const mockRelease = { _id: 'release' }

    releaseModelMocks.findOne.mockResolvedValue(mockRelease)

    vi.mocked(authorisation.release).mockImplementation(async (_user, _model, action, _release) => {
      if (action === ReleaseAction.View) return { success: true, id: '' }
      if (action === ReleaseAction.Delete)
        return { success: false, info: 'You do not have permission to delete this release.', id: '' }

      return { success: false, info: 'Unknown action.', id: '' }
    })

    expect(() => deleteRelease({} as any, 'test', 'test')).rejects.toThrowError(
      /^You do not have permission to delete this release./,
    )
    expect(releaseModelMocks.save).not.toBeCalled()
  })

  test('deleteRelease > should throw a bad req when attempting to delete a release from a mirrored model', async () => {
    const mockRelease = { _id: 'release' }

    releaseModelMocks.findOne.mockResolvedValue(mockRelease)

    vi.mocked(authorisation.release).mockImplementation(async (_user, _model, action, _release) => {
      if (action === ReleaseAction.View) return { success: true, id: '' }
      if (action === ReleaseAction.Delete)
        return { success: false, info: 'Cannot delete a file from a mirrored model.', id: '' }

      return { success: false, info: 'Unknown action.', id: '' }
    })

    expect(() => deleteRelease({} as any, 'test', 'test')).rejects.toThrowError(
      /^Cannot delete a file from a mirrored model./,
    )
    expect(releaseModelMocks.save).not.toBeCalled()
  })

  test('removeFileFromReleases > no permission', async () => {
    const mockUser: any = { dn: 'test' }
    const mockModel: any = { id: 'test', settings: { mirror: { sourceModelId: '' } } }
    const mockRelease = { _id: 'release' }

    vi.mocked(authorisation.releases).mockResolvedValue([
      {
        success: false,
        info: 'You do not have permission to update these releases.',
        id: '',
      },
    ])

    releaseModelMocks.find.mockResolvedValueOnce([mockRelease, mockRelease])

    const result = removeFileFromReleases(mockUser, mockModel, '')

    expect(result).rejects.toThrowError(/^You do not have permission to update these releases./)
    expect(releaseModelMocks.updateMany).not.toBeCalled()
  })

  test('removeFileFromReleases > success', async () => {
    const mockUser: any = { dn: 'test' }
    const mockModel: any = { id: 'test', settings: { mirror: { sourceModelId: '' } } }
    const mockRelease = { _id: 'release' }
    const resultObject = { modifiedCount: 2, matchedCount: 2 }

    releaseModelMocks.find.mockResolvedValueOnce([mockRelease, mockRelease])
    releaseModelMocks.updateMany.mockResolvedValueOnce(resultObject)

    const result = await removeFileFromReleases(mockUser, mockModel, '')

    expect(result).toEqual(resultObject)
  })

  test('removeFileFromReleases > should throw a bad req when attempting to remove a file from a mirrored model', async () => {
    const mockUser: any = { dn: 'test' }
    const mockModel: any = { id: 'test', settings: { mirror: { sourceModelId: '123' } } }
    const mockRelease = { _id: 'release' }

    vi.mocked(authorisation.releases).mockResolvedValue([
      {
        success: false,
        info: 'Cannot remove a file from a mirrored model.',
        id: '',
      },
    ])

    releaseModelMocks.find.mockResolvedValueOnce([mockRelease, mockRelease])

    expect(() => removeFileFromReleases(mockUser, mockModel, '123')).rejects.toThrowError(
      /^Cannot remove a file from a mirrored model./,
    )
    expect(releaseModelMocks.delete).not.toBeCalled()
  })

  test('getFileByReleaseFileName > success', async () => {
    const mockUser: any = { dn: 'test' }
    const modelId = 'example'
    const semver = '1.0.0'
    const fileName = 'test.png'

    fileMocks.getFilesByIds.mockResolvedValueOnce([{ name: 'test.png' }])

    const file = await getFileByReleaseFileName(mockUser, modelId, semver, fileName)

    expect(file.name).toBe('test.png')
  })

  test('getFileByReleaseFileName > file not found', async () => {
    const mockUser: any = { dn: 'test' }
    const modelId = 'example'
    const semver = '1.0.0'
    const fileName = 'test.png'

    fileMocks.getFilesByIds.mockResolvedValueOnce([{ name: 'not_test.png' }])

    const result = getFileByReleaseFileName(mockUser, modelId, semver, fileName)
    expect(result).rejects.toThrowError(/^The requested file name was not found on the release./)
  })

  test('getReleasesForExport > release not found', async () => {
    const mockUser: any = { dn: 'test' }
    const modelId = 'example'
    const semvers = ['1.0.0']
    releaseModelMocks.find.mockResolvedValueOnce([])

    const result = getReleasesForExport(mockUser, modelId, semvers)

    expect(result).rejects.toThrowError(/^The following releases were not found./)
  })

  test('getReleasesForExport > release not found', async () => {
    const mockUser: any = { dn: 'test' }
    const modelId = 'example'
    const semvers = ['1.0.0']
    releaseModelMocks.find.mockResolvedValueOnce([{ semver: '1.0.0' }])
    vi.mocked(authorisation.releases).mockResolvedValue([
      { info: 'You do not have permission', success: false, id: '' },
    ])

    const result = getReleasesForExport(mockUser, modelId, semvers)

    expect(result).rejects.toThrowError(/^You do not have the necessary permissions to export these releases./)
  })

  test('getReleasesForExport > return releases', async () => {
    const mockUser: any = { dn: 'test' }
    const modelId = 'example'
    const semvers = ['1.0.0']
    const releases = [{ semver: '1.0.0' }]
    releaseModelMocks.find.mockResolvedValueOnce(releases)

    const result = await getReleasesForExport(mockUser, modelId, semvers)

    expect(result).toBe(releases)
  })

  test('getTotalFileSize > returns fileIds', async () => {
    releaseModelMocks.group.mockResolvedValueOnce([{ fileIds: [42] }])
    const size = await getAllFileIds('example', ['1', '2', '3'])

    expect(size).toStrictEqual([42])
  })

  test('getTotalFileSize > returns empty list if no file Ids', async () => {
    releaseModelMocks.group.mockResolvedValueOnce([])
    const size = await getAllFileIds('example', ['1', '2', '3'])

    expect(size).toStrictEqual([])
  })
})
