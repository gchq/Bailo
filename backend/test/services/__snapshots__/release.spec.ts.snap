// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`services > release > getModelReleases > convertSemverQueryToMongoQuery functions 1`] = `
[
  {
    "$and": [
      {
        "$or": [
          {
            "semver.major": {
              "$gte": 2,
            },
            "semver.minor": {
              "$gte": 2,
            },
            "semver.patch": {
              "$gte": 0,
            },
          },
          {
            "semver.major": {
              "$gt": 2,
            },
          },
          {
            "semver.major": {
              "$gte": 2,
            },
            "semver.minor": {
              "$gt": 2,
            },
          },
        ],
      },
      {
        "$or": [
          {
            "semver.major": {
              "$lte": 2,
            },
            "semver.minor": {
              "$lte": 3,
            },
            "semver.patch": {
              "$lt": 0,
            },
          },
          {
            "semver.major": {
              "$lt": 2,
            },
          },
          {
            "semver.major": {
              "$lte": 2,
            },
            "semver.minor": {
              "$lt": 3,
            },
          },
        ],
      },
    ],
    "modelId": "modelId",
  },
]
`;

exports[`services > release > getModelReleases > convertSemverQueryToMongoQuery functions with less than 1`] = `
[
  {
    "$and": [
      {
        "$or": [
          {
            "semver.major": {
              "$lte": 2,
            },
            "semver.minor": {
              "$lte": 2,
            },
            "semver.patch": {
              "$lt": 2,
            },
          },
          {
            "semver.major": {
              "$lt": 2,
            },
          },
          {
            "semver.major": {
              "$lte": 2,
            },
            "semver.minor": {
              "$lt": 2,
            },
          },
        ],
      },
    ],
    "modelId": "modelID",
  },
]
`;

exports[`services > release > getModelReleases > good 1`] = `
[
  {
    "modelId": "modelId",
  },
]
`;

exports[`services > release > getModelReleases > good 2`] = `
[
  {
    "updatedAt": -1,
  },
]
`;

exports[`services > release > getModelReleases > good 3`] = `
[
  {
    "as": "model",
    "foreignField": "id",
    "from": "v2_models",
    "localField": "modelId",
  },
]
`;

exports[`services > release > getModelReleases > good 4`] = `
[
  {
    "$set": {
      "model": {
        "$arrayElemAt": [
          "$model",
          0,
        ],
      },
    },
  },
]
`;
